### PR TITLE
Add Non-nilable map using accessor functions

### PR DESCRIPTION
**Pull request by daviditen**
Update map to handle non-nillable shared classes. Use a nilable backing array and use accessor functions for accessing the values in different ways.

### CODE REVIEW

File: modules/standard/Map.chpl [hunk 1]
```diff
@@ -254,6 +257,70 @@ module Map {
...
+    pragma "no doc"
+    proc const this(k: keyType)
+    where isNonNilableClass(valType) {
+      compilerError("Cannot access nilable class directly. Use an",
+                    " appropriate accessor method instead.");
+    }
+
+    /* Get a borrowed reference to the element at position `k`.
+     */
+    proc getBorrowed(k: keyType) {
```
Comment (dlongnecke-cray) [inline modules/standard/Map.chpl]: What about if the value type is not a class? Does it make sense to allow a call to `getBorrowed`? Can we restrict this with something like `where isClass(valType)`?

File: modules/standard/Map.chpl [hunk 2]
```diff
@@ -109,6 +102,17 @@ module Map {
       this.parSafe = parSafe;
     }
 
+    proc init(type keyType, type valType, param parSafe=false)
+    where isNonNilableClass(valType) {
+      if isOwnedClass(valType) {
+        compilerError("maps of non-nilable owned values are not allowed");
```
[the remark this answers is not in the document]
Comment (daviditen) [reply modules/standard/Map.chpl]: They get errors about needing to specify lifetimes. I can add an error here for them.

File: modules/standard/Map.chpl [hunk 3]
```diff
@@ -289,17 +356,36 @@ module Map {
       }
     }
 
+    pragma "no doc"
+    iter items() const ref where isNonNilableClass(valType) {
+      try! {
+        for key in myKeys {
+          yield (key, vals[key]: valType);
```
[the remark this answers is not in the document]
Comment (daviditen) [reply modules/standard/Map.chpl]: I don't think the return value can be a tuple, only something that `isNonNilableClass` due to the `where` constraint.

File: modules/standard/Map.chpl [hunk 4]
```diff
@@ -382,6 +469,18 @@ module Map {
       return true;
     }
 
+    /* If the map doesn't contain a value at position `k` add one and
+       set it to `v`. If the map already contains a value at position
+       `k`, update it to the value `v`.
+     */
+    proc addOrSet(k: keyType, in v: valType) {
```
[the remark this answers is not in the document]
Comment (daviditen) [reply modules/standard/Map.chpl]: Works for me in both spots.

File: modules/standard/Map.chpl (hunk 1 above)
Comment (daviditen) [reply modules/standard/Map.chpl]: Sure, that where clause sounds reasonable to me.

File: modules/standard/Map.chpl (hunk 3 above)
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply modules/standard/Map.chpl]: It looks like it's yielding a tuple of `(key, vals[key]: valType)`, though? I'm worried about whether it's possible for `const ref` to introduce the "returning a reference to a temporary" problem we experienced earlier with casts.

File: modules/standard/Map.chpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply modules/standard/Map.chpl]: Why don't we tentatively go with _yes_ to adding `in` intent to `k` for `add` and `addOrSet` for now.

File: modules/standard/Map.chpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply modules/standard/Map.chpl]: Sounds good.

[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/Map.chpl]: > They get errors about needing to specify lifetimes. I can add an error here for them.

Did you try specifying the lifetimes?

File: modules/standard/Map.chpl [hunk 5]
```diff
@@ -341,7 +427,7 @@ module Map {
                `false` otherwise.
      :rtype: bool
     */
-    proc add(k: keyType, v: valType): bool {
+    proc add(k: keyType, in v: valType): bool {
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/Map.chpl]: If the common situation is that the value `k` is stored in some data structure here then `in` intent would make sense

File: modules/standard/Map.chpl (hunk 3 above)
[the remark this answers is not in the document]
Comment (daviditen) [reply modules/standard/Map.chpl]: Oh, right. I'm pretty sure valgrind would catch that, and it didn't complain about this when I checked.

File: modules/standard/Map.chpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (daviditen) [reply modules/standard/Map.chpl]: > Did you try specifying the lifetimes?

I did not.

File: modules/standard/Map.chpl (hunk 3 above)
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/Map.chpl]: Iterators are pretty different in implementation - and the differences make this more likely to be OK. But it's unclear to me if we ought to be issuing compilation errors for such a pattern. If it works with valgrind I'd say that's good enough for today.

**Pull request merged**